### PR TITLE
STEP files support

### PR DESCRIPTION
**Pull request by martinRenou**

### CODE REVIEW

**Comment by trungleduc** [comment]
you can change `"build"` to `"build_dir"` here to fix the release action

https://github.com/jupytercad/jupytercad/blob/f980c3f1196d24154038f672fbee2e86f4bf66ab/python/jupytercad_app/webpack.config.js#L32C7-L32C12

**Comment by martinRenou** [comment]
please update snapshots

File: packages/schema/src/model.ts [hunk 1]
```diff
@@ -300,9 +300,11 @@ export class JupyterCadDoc
   }
 
   get objects(): Array<IJCadObject> {
-    return this._objects.map(
+    const objs = this._objects.map(
       obj => JSONExt.deepCopy(obj.toJSON()) as IJCadObject
     );
+    console.log(objs);
```
Comment (trungleduc) [inline packages/schema/src/model.ts:306]:
```suggestion
```

File: packages/schema/src/model.ts [hunk 2]
```diff
@@ -483,6 +485,76 @@ export class JupyterCadDoc
...
+  }
+
+  get objectsChanged(): ISignal<IJupyterCadDoc, IJcadObjectDocChange> {
+    return this._objectChanged;
+  }
+
+  get objects(): Array<IJCadObject> {
+    const source = this._source.toJSON();
+
+    console.log('Source is', this._source, source);
```
Comment (trungleduc) [inline packages/schema/src/model.ts:510]:
```suggestion
```

File: packages/schema/src/model.ts [hunk 3]
```diff
@@ -483,6 +485,76 @@ export class JupyterCadDoc
...
+
+    if (!source) {
+      return [];
+    }
+
+    return [
+      {
+        name: 'Step File', // TODO get file name?
```
Comment (trungleduc) [inline packages/schema/src/model.ts:518]: Yes, i think the file name is better

File: packages/schema/src/model.ts [hunk 4]
```diff
@@ -483,6 +485,76 @@ export class JupyterCadDoc
...
+  private _sourceObserver = (events: Y.YEvent<any>[]): void => {
+    const changes: Array<{
+      name: string;
+      key: keyof IJCadObject;
+      newValue: IJCadObject;
+    }> = [];
+    events.forEach(event => {
+      event.keys.forEach((change, key) => {
+        changes.push({
+          name: 'Step File',
```
Comment (trungleduc) [inline packages/schema/src/model.ts:542]: Same here

File: packages/schema/src/schema/any.json [hunk 5]
```diff
@@ -2,12 +2,16 @@
...
   "additionalProperties": false,
   "properties": {
-    "Shape": {
+    "Content": {
       "type": "string",
-      "description": "The BREP string of the shape"
+      "description": "The string content of the object"
+    },
+    "Type": {
+      "type": "string",
```
Comment (trungleduc) [inline packages/schema/src/schema/any.json:13]: I think an enum of supported type is better, it helps validate the JCAD file and generate the correct interface for `_loadObjectFile` function

File: packages/occ-worker/src/occapi.ts [hunk 6]
```diff
@@ -349,6 +347,40 @@ export function _loadBrep(arg: {
...
+    return shape;
+  } else {
+    console.error('Something in OCCT went wrong trying to read');
+    return undefined;
+  }
+}
+
+export function _loadObjectFile(arg: {
+  content: string;
+  type: string;
```
Comment (trungleduc) [inline packages/occ-worker/src/occapi.ts:372]: a more specific interface in form of `step | brep` would be better. I goes with the comment in the `any.json` schema.

File: python/jupytercad_core/src/stepplugin/modelfactory.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
+import { IJupyterCadDoc, JupyterCadModel } from '@jupytercad/schema';
+import { DocumentRegistry } from '@jupyterlab/docregistry';
+import { Contents } from '@jupyterlab/services';
+
+/**
+ * A Model factory to create new instances of JupyterCadModel.
+ */
+export class JupyterCadJcadStepModelFactory
```
Comment (trungleduc) [inline python/jupytercad_core/src/stepplugin/modelfactory.ts:8]:
```suggestion
export class JupyterCadStepModelFactory
```

**Review by trungleduc — CHANGES_REQUESTED** [review]

File: packages/schema/src/model.ts (hunk 3 above)
Comment (martinRenou) [reply packages/schema/src/model.ts:515]: Unfortunately the ydoc does not known anything about the name of the file it originated from, or am I missing something?

Comment (trungleduc) [reply packages/schema/src/model.ts:515]: ah indeed. So let's keep it like this.

**Comment by trungleduc** [comment]
Thanks @martinRenou!

**Pull request merged**